service/backend/pchain/account.go

```diff
@@ -162,6 +169,98 @@ func (b *Backend) AccountCoins(ctx context.Context, req *types.AccountCoinsReque
+	addr, err := address.ParseToID(req.AccountIdentifier.Address)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInvalidInput, "malformed address")
+	}
+
+	metadata := req.AccountIdentifier.Metadata
+	var validatorNodeID ids.NodeID
+	var nodeIDs []ids.NodeID
+	if metadata != nil && metadata[pmapper.MetadataValidatorNodeID] != nil {
+		nodeIDStr, ok := metadata[pmapper.MetadataValidatorNodeID].(string)
+		if !ok {
+			return nil, service.WrapError(service.ErrInvalidInput, "validator_node_id must be a string")
+		}
+
+		validatorNodeID, err = ids.NodeIDFromString(nodeIDStr)
+		if err != nil {
+			return nil, service.WrapError(service.ErrInvalidInput, "malformed validator_node_id ")
+		}
+
+		nodeIDs = append(nodeIDs, validatorNodeID)
+	}
+
+	validators, err := b.pClient.GetCurrentValidators(ctx, constants.PrimaryNetworkSubnetID, nodeIDs)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInternalError, "unable to fetch validators")
+	}
+
+	validatorRewards := big.NewInt(0)
+	delegationFeeRewards := big.NewInt(0)
+	delegationRewards := big.NewInt(0)
+
+	for _, v := range validators {
+		isValidatorOwner := false
+
+		if slices.Contains(v.ValidationRewardOwner.Addresses, addr) {
+			validatorRewards.Add(validatorRewards, new(big.Int).SetUint64(*v.PotentialReward))
+			isValidatorOwner = true
+		}
+
+		for _, d := range v.Delegators {
+			delegationReward := new(big.Int).SetUint64(*d.PotentialReward)
+			// Representing delegation fee as percentage of 10^6, the same way as
+			// DelegatorShare parameter of AddValidator transaction
+			delegationFeePct := big.NewInt(int64(v.DelegationFee * 10_000))
+			delegationFee := new(big.Int).Div(new(big.Int).Mul(delegationReward, delegationFeePct), big.NewInt(1_000_000))
+
+			// if the address we are searching is the validator owner, add the delegation fee
+			if isValidatorOwner {
+				delegationFeeRewards.Add(delegationFeeRewards, delegationFee)
+			}
+
+			// ff the address delegated to the current validator, add the potential reward minus fee
+			if slices.Contains(d.RewardOwner.Addresses, addr) {
+				delegateReward := new(big.Int).Sub(delegationReward, delegationFee)
+				delegationRewards.Add(delegationRewards, delegateReward)
+			}
+		}
+	}
+
+	totalRewards := big.NewInt(0)
+	totalRewards.Add(totalRewards, validatorRewards)
+	totalRewards.Add(totalRewards, delegationFeeRewards)
+	totalRewards.Add(totalRewards, delegationRewards)
+
+	height, err := b.pClient.GetHeight(ctx)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInternalError, "unable to get block height")
+	}
+
+	block, err := b.indexerParser.ParseNonGenesisBlock(ctx, "", height)
+	if err != nil {
+		return nil, service.WrapError(service.ErrInternalError, "unable to get current block")
+	}
+
+	return &types.AccountBalanceResponse{
+		BlockIdentifier: &types.BlockIdentifier{
+			Index: int64(height),
+			Hash:  block.BlockID.String(),
+		},
```

Comment:
@bilgehansahin this is an attempt to associate an height to the calculated rewards right? I think we do something similar in the codebase.

Reply:
Yes, we do that with other /account/balance and /account/coins calls.